src/goto-instrument/contracts/instrument_spec_assigns.cpp

```diff
@@ -91,30 +94,119 @@ void instrument_spec_assignst::check_inclusion_assignment(
 
 void instrument_spec_assignst::track_static_locals(goto_programt &dest)
 {
-  auto call_graph =
-    call_grapht::create_from_root_function(functions, function_id, true)
-      .get_directed_graph();
+  covered_linest covered_lines;
+  const auto &found = functions.function_map.find(function_id);
+  INVARIANT(
+    found != functions.function_map.end(),
+    "the instrumented function must exist in the model");
+
+  const goto_programt &body = found->second.body;
+  traverse_instructions(
+    body.instructions.begin(), body.instructions.end(), covered_lines);
+
+  std::unordered_set<symbol_exprt, irep_hash> symbols;
+  collect_static_symbols(covered_lines, symbols);
 
-  for(const auto &sym_pair : st)
+  for(const auto &sym : symbols)
+    create_snapshot(create_car_from_stack_alloc(sym), dest);
+}
+
+size_t line_to_size_t(const irep_idt &line)
+{
+  PRECONDITION(!line.empty());
+  std::stringstream stream(id2string(line));
+  size_t res;
+  stream >> res;
+  return res;
+}
+
+void instrument_spec_assignst::traverse_instructions(
+  const goto_programt::const_targett &begin,
+  const goto_programt::const_targett &end,
+  covered_linest &covered_lines) const
+{
+  goto_programt::const_targett it(begin);
+
+  while(it != end)
   {
-    const auto &sym = sym_pair.second;
-    if(sym.is_static_lifetime)
+    const auto &loc = it->source_location();
+    const auto &loc_fun = loc.get_function();
+    if(!loc_fun.empty())
     {
-      auto fname = sym.location.get_function();
-      if(
-        !fname.empty() &&
-        (fname == function_id || call_graph.get_node_index(fname).has_value()))
+      auto &itv = covered_lines[loc_fun];
+      const auto &loc_line = loc.get_line();
+      if(!loc_line.empty())
+        itv.update(line_to_size_t(loc_line));
+    }
+
+    if(it->is_function_call())
+    {
+      const auto &fun_expr = it->call_function();
+
+      PRECONDITION_WITH_DIAGNOSTICS(
+        fun_expr.id() == ID_symbol,
+        "local static search requires function pointer removal");
+      const irep_idt &fun_id = to_symbol_expr(fun_expr).get_identifier();
+
+      const auto &found = functions.function_map.find(fun_id);
+
+      if(found != functions.function_map.end())
```

Comment:
This should be an `INVARIANT`: all functions that are called must exist. They might not have a body, but AFAIK they must exist.